[OG-135] Course groups

LisieckaK00 opened the pull request:
This PR aim was to implement backend for course groups view. That includes students activity, profile, equipment and chests and also opening chest (#202)

krudny commented:
Please take into account that issue you will create is automatically assigned to you. If critical (criteria are essential for us so probably yes) then deadline is to friday/max sunday as agreed with our supervisor.

kligarski commented on polymorphia-frontend/components/equipment/modals/opening-chest/OpeningChestModalContent.tsx:
```diff
@@ -32,12 +39,20 @@ export default function OpeningChestModalContent({
   const isSm = useMediaQuery({ maxWidth: 768 });
   const areAllItemsOverLimit =
     equipment.base.chestItems?.every((item) => item.isLimitReached) ?? false;
+  // not using useTargetContext() - if we're in target context we're using it to fetch data for a particular student, otherwise we're using currently logged student's data
```
For me, solution with `useContext` makes sense, it's just kind of reversed - if `OpeningChestModal` is in `TargetContext`, then it adapts to use `targetId` from that context. If the modal is not in `TargetContext`, it's used as standalone component and it should use current user's (student's) ID. This is the `magical sometimes` you mentioned.

I'd just make it clearer:

```tsx
export default function useStudentChests(
  overrideStudentId: number | null
): UseStudentChest {
  const { courseId } = useUserDetails();
  const { userRole } = useUserContext();
  const { data, isLoading, isError, refetch } = useQuery({
    queryKey: ["studentChest", overrideStudentId],
    queryFn: () => CourseGroupsService.getStudentChests(courseId, overrideStudentId),
    refetchOnWindowFocus: false,
    enabled:
      overrideStudentId !== null || userRole === Roles.STUDENT, // null for student is acceptable (it means: use current user)
  });

  return { data, isLoading, isError, refetch };
}
```

and then handle undefined `overrideStudentId` gracefully in `CourseGroupsService.getStudentChests` by not passing that query param if `overrideStudentId` is `null`.

Approach with prop drilling is also okay for me.

kligarski commented on polymorphia-backend/src/main/java/com/agh/polymorphia_backend/service/course/reward/AssignedRewardService.java:
```diff
@@ -77,14 +73,8 @@ public Map<Long, BigDecimal> getTotalBonusByEventSection(List<AssignedItem> assi
-                .findFirst()
+    public AssignedChest getAssignedChest(Long assignedChestId) {
+        return assignedChestRepository.findNotUsedAssignedChestsById(assignedChestId)
                 .orElseThrow(() -> new ResponseStatusException(HttpStatus.BAD_REQUEST, "Nie znaleziono przypisanej skrzynki."));
```
```suggestion
    public AssignedChest getUnopenedAssignedChest(Long assignedChestId) {
        return assignedChestRepository.findNotUsedAssignedChestsById(assignedChestId)
.orElseThrow(() -> new ResponseStatusException(HttpStatus.BAD_REQUEST, "Nie znaleziono nieotwartej skrzynki, przypisanej do zwierzaka, o podanym ID."));
```

kligarski commented on polymorphia-frontend/hooks/course/useStudentSummary/index.tsx:
```diff
@@ -1,13 +1,13 @@
-    queryFn: () => CourseGroupsService.getStudentSummary(userId!),
+    queryFn: () => CourseGroupsService.getStudentSummary(courseId, userId),
     refetchOnWindowFocus: false,
     enabled: userId !== null,
```
This seems to be redundant now?

LisieckaK00 commented:
I changed the title, as per my last comment I added student activities in this PR (the middle column) to avoid resolving conflicts in a few PRs.

So not to block this PR, I'll leave these native queries for now and will come back for them in a new one.

I made some updates (mostly context logic), will come back soon with develop merge and profile update.

kligarski commented on polymorphia-backend/src/main/java/com/agh/polymorphia_backend/repository/course/reward/assigned/AssignedChestRepository.java:
```diff
@@ -2,31 +2,45 @@
 
+    @Lock(LockModeType.PESSIMISTIC_WRITE)
+    @Query(value = SELECT_ASSIGNED_CHEST_NOT_USED)
+    Optional<AssignedChest> findNotUsedAssignedChestsByIdWithLock(@Param("assignedChestId") Long assignedChestId);
```
I think that @krudny used DB query to also validate whether assignedChestId belongs to the current user's animal:

```java
Long animalId = animalService.validateAndGetAnimalId(courseId);
AssignedChest assignedChest = assignedRewardService.getAssignedChestByIdAndAnimalIdWithLock(
    requestDto.getAssignedChestId(), animalId);
```

It seems like we're missing this logic now unless I missed something.

LisieckaK00 replied:
I'm validating that in a different place - I'm taking the `animalId` from `assignedChest` and then checking in `EquipmentService::validatePermissions(courseId, animalId)` the permissions (if student - then if it's his chest, otherwise if instructor/coordinator has permissions for given student)

kligarski replied:
Okay, thank you for the explanation.

kligarski commented on polymorphia-backend/src/main/java/com/agh/polymorphia_backend/repository/gradable_event/GradeRepository.java:
```diff
@@ -21,4 +21,21 @@ public interface GradeRepository extends JpaRepository<Grade, Long> {
+    g.animal_id as animalName,
+    ge.id as id,
+    ge.name as gradableEventName,
+    cg.xp as gainedXp,
```
Gradable events can have multiple criteria -> multiple criteria grades. We don't aggregate them in this query which is incorrect.

To data from staging db (at least my most recent dump), add the second row:

<img width="737" height="126" alt="Image" src="https://github.com/user-attachments/assets/3f6e3b85-cb1e-41be-a5ef-23018ce828ee" />

Then `Projekt` is repeated.

<img width="1467" height="702" alt="Image" src="https://github.com/user-attachments/assets/f67824c7-22a6-444a-866f-10166b781226" />

kligarski reviewed (commented):
Additional comments:

We should invalidate course groups target list when chest is opened (if xp has changed after opening the chest, old xp is still visible in target list).

Also, student summary should be invalidated as well because the evolution stage can change after opening the chest.

LisieckaK00 commented on polymorphia-frontend/hooks/course/useStudentLastActivity/index.tsx:
```diff
@@ -1,13 +1,16 @@
     queryKey: ["studentLastActivity", userId],
-    queryFn: () => CourseGroupsService.getStudentLastActivity(userId),
+    queryFn: () => CourseGroupsService.getStudentLastActivity(userId, courseId),
+    enabled: userId !== null,
```
The `targetId` has a type `number | null` so at the begging in `CourseGroupsProvider` the student summary and activity are called with `userId` being null. What should I do to avoid it? It is in the provider, so I'm not sure how this should be resolved :c

kligarski replied:
It's okay that it can be null but we shouldn't be declaring that it isn't in hook's `userId` argument and then forcing possibly null value with `!` operator. Types are meant to help us.

Just add nullability to property in hook, e.g.

```tsx
export default function useStudentLastActivity(
  userId: number | null // add nullability here
): UseStudentLastActivity {
```

Same with summary. Then remove `!` operator in places where the hooks are used.

LisieckaK00 replied:
sure, thank you! 🫡

krudny commented:
<img width="181" height="226" alt="image" src="https://github.com/user-attachments/assets/94c8a748-add1-444b-8cbe-42934f471ca8" />

i think that it is not db case

kligarski commented on polymorphia-frontend/components/equipment/modals/opening-chest/types.ts:
```diff
@@ -3,4 +3,5 @@ import { EquipmentChestResponseDTO } from "@/interfaces/api/equipment";
 export interface OpeningChestModalProps {
   equipment: EquipmentChestResponseDTO;
   onClose: () => void;
+  targetStudentIdOverride?: number | null;
```
Target context uses `number | null`. We're passing the prop directly (which is what we should do) so it's okay.

LisieckaK00 commented on polymorphia-frontend/hooks/course/useStudentItems/index.tsx:
```diff
@@ -1,15 +1,23 @@
 export default function useStudentItems(
   userId: number | null
 ): UseStudentItems {
+  const { courseId } = useUserDetails();
```
It doesn't go through 5 components. It's used only in  `StudentInfo()`:
```ts
const { selectedTarget, targetId } = useTargetContext();
  const {
    data: items,
    isLoading: isItemsLoading,
    isError: isItemsError,
  } = useStudentItems(targetId);
  const {
    data: chests,
    isLoading: isChestsLoading,
    isError: isChestsError,
  } = useStudentChests(targetId);
```
So I don't see much gain by getting that inside the methods, but let me know if I should update it

LisieckaK00 replied:
Makes sense. Should all components like student activity, profile, chests and items take no argument and pull `targetId` from target context for consistency? Or `userId` (/`targetId`) should be but just made optional?

LisieckaK00 replied:
But now i was using in the opening modal the `targetId` for invalidating queries for given target:
```ts
      await queryClient.invalidateQueries({
        queryKey: ["studentItems", targetId],
      });
      await queryClient.invalidateQueries({
        queryKey: ["studentSummary", targetId],
      });
```
should I be doing now:
```ts
      await queryClient.invalidateQueries({
        queryKey: ["studentItems"],
      });
      await queryClient.invalidateQueries({
        queryKey: ["studentSummary"],
      });
```
? But we'll be invalidating more than necessary

LisieckaK00 commented:
I've merged the latest changes from `develop` and tested the flow using example data (including the grading process and the group view). Everything appears to be working correctly on my end.
However, I'd appreciate it if you could take another look at the code and test it as well, just to make sure everything behaves as expected.
